src/RecommendedOptions.cmake

```
@@ -0,0 +1,136 @@
+set(options
+    "ENABLE_CACHE\;${MAKEFILE_OR_NINJA}\;${MAKEFILE_OR_NINJA}\;Enable ccache on Unix"
+    "WARNINGS_AS_ERRORS\;OFF\;ON\;Treat warnings as Errors"
+    "ENABLE_CLANG_TIDY\;OFF\;${MAKEFILE_OR_NINJA}\;Enable clang-tidy analysis during compilation"
+    "ENABLE_CONAN\;ON\;ON\;Automatically integrate Conan for package management"
+    "ENABLE_COVERAGE\;OFF\;OFF\;Analyze and report on coverage"
+    "ENABLE_SANITIZER_ADDRESS\;OFF\;ON\;Make memory errors into hard runtime errors (windows/linux/macos)"
+    "ENABLE_SANITIZER_UNDEFINED_BEHAVIOR\;OFF\;${GCC_OR_CLANG}\;Make certain types (numeric mostly) of undefined behavior into runtime errors"
+    "ENABLE_CPPCHECK\;OFF\;${MAKEFILE_OR_NINJA}\;Enable cppcheck analysis during compilation"
+    "ENABLE_IPO\;OFF\;OFF\;Enable whole-program optimization"
+    "ENABLE_INCLUDE_WHAT_YOU_USE\;OFF\;OFF\;Enable include-what-you-use analysis during compilation"
+    "ENABLE_PCH\;ON\;OFF\;Enable pre-compiled-headers support"
+    "ENABLE_DOXYGEN\;OFF\;OFF\;Build documentation with Doxygen"
+    "ENABLE_USER_LINKER\;OFF\;OFF\;Allow custom linker settings"
+    "ENABLE_BUILD_WITH_TIME_TRACE\;OFF\;OFF\;Generates report of where compile-time is spent"
+    "ENABLE_UNITY\;OFF\;OFF\;Merge C++ files into larger C++ files, can speed up compilation sometimes"
+    "ENABLE_SANITIZER_LEAK\;OFF\;OFF\;Make memory leaks into hard runtime errors"
+    "ENABLE_SANITIZER_THREAD\;OFF\;OFF\;Make thread race conditions into hard runtime errors"
+    "ENABLE_SANITIZER_MEMORY\;OFF\;OFF\;Make other memory errors into runtime errors")
+
+foreach(option ${options})
+  list(
+    GET
+    option
+    0
+    option_name)
+  list(
+    GET
+    option
+    1
+    option_user_default)
+  list(
+    GET
+    option
+    2
+    option_developer_default)
+  list(
+    GET
+    option
+    3
+    option_description)
+
+  if(DEFINED ${option_name}_DEFAULT)
+    if(DEFINED ${option_name}_DEVELOPER_DEFAULT OR DEFINED ${option_name}_USER_DEFAULT)
+      message(
+        SEND_ERROR
+          "You have separately defined user/developer defaults and general defaults for ${option_name}. Please either provide a general default OR separate developer/user overrides"
+      )
+    endif()
+
+    set(option_user_default ${${option_name}_DEFAULT})
+    set(option_developer_default ${${option_name}_DEFAULT})
+  endif()
+
+  if(DEFINED ${option_name}_USER_DEFAULT)
+    set(option_user_default ${${option_name}_USER_DEFAULT})
+  endif()
+
+  if(DEFINED ${option_name}_DEVELOPER_DEFAULT)
+    set(option_developer_default ${${option_name}_DEVELOPER_DEFAULT})
+  endif()
+
+  cmake_dependent_option(
+    OPT_${option_name}
+    "${option_description}"
+    ${option_developer_default}
+    ENABLE_DEVELOPER_MODE
+    ${option_user_default})
+
+  if(OPT_${option_name})
+    set(${option_name}_VALUE ${option_name})
+  else()
+    unset(${option_name}_VALUE)
+  endif()
+endforeach()
+
+macro(recommended_project_options)
```

Should we change this to `dynamic_project_options`? This implies that `project_options` is not recommended or has bad defaults! 

```suggestion
macro(dynamic_project_options)
```

---

It sets the recommended defaults, which is to have conan enabled. And it actually does make sense to allow the developer to disable conan, if they want to run it manually, which, IIRC is the entire reason we moved to a conanfile.txt.

How about `recommended_default_project_options`

---

> It sets the recommended defaults

And also adds dynamic global options for all the function parameters that can be changed from CLI or GUI. That's the part I have problems with. We could instead use local variables, which does not affect the reproducibility.

> And it actually does make sense to allow the developer to disable conan, if they want to run it manually, which, IIRC is the entire reason we moved to a conanfile.txt.

As far as I understand, a project either uses Conan or does not. If it does, then the option should be written in the CMake file. The reason for using conanfile.txt is that it is "declarative" like all the package managers (e.g. cargo, npm, etc).

---

> As far as I understand, a project either uses Conan or does not. If it does, then the option should be written in the CMake file. The reason for using conanfile.txt is that it is "declarative" like all the package managers (e.g. cargo, npm, etc).

Incorrect. We are using the cmake package finders. Those packages can come from

 * conan run automatically by cmake
 * conan run manually by the user
 * system provided packages

---

> And also adds dynamic global options for all the function parameters that can be changed from CLI or GUI. That's the part I have problems with. We could instead use local variables, which does not affect the reproducibility.

I do not see how this affects reproducibility. It gives developers flexibility to change the options they need during the development process. None of these options fundamentally change the nature of your C++ project.

They also allow a user of the program to trivially easily test building on an unsupported compiler by setting the high level option of `DEVELOPER_MODE` to `Off`. This helps in the cases where an untested compiler introduces different warnings, and does not force the user to manually edit the CMakeLists or fix the warnings before they can use the application.

---

Could you update the readme example, so we can see how this comes together?